Add more help for "what do do when reading a table fails"

hamogu opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
Add a gallery of examples to guide users from "oh my table does not read" to "maybe try this option" to eventually writing specific Reader/Writer classes.
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->


<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #6365

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will
review this pull request of some common things to look for. This list
is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] If the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

hamogu commented:
@taldcroft I would like your early feedback if you think such a doc addition is useful. I occasionally tell people to do the things I write down here, so I thought it might be useful to add that to the official documentation. I've been writing for a while now, but I'd like your opinion before I go and make sure that all examples work in sphinx etc. (That's why this is still marked as a draft.)
The style is less technical than most of the io.ascii documentation, so you might suggest to put this into astropy learn instead. In that case, I don't want to waste my time getting all the intersphinx right.

On the other hand, my impression is that more people read (or are directed by Google to) the actual astropy documentation, which is why I would like it in there instead of as a "code snippet" on astropy learn or a similar forum.

taldcroft commented:
I am happy to have the examples live in the astropy docs, and I agree with your reasoning about discovery. There is already precedent in the fixed-width gallery.

pllim commented:
Maybe we should loop in the Learn team... @kelle @eblur @adrn et al.

kelle commented:
The Learn team, and @jonathansick in particular, is working on building infrastructure to auto-populate an example gallery from the examples in the Sphinx docs using those `EXAMPLE START` and `EXAMPLE END` tags. So this looks great to us!

taldcroft commented on docs/io/ascii/howtoexamples.rst:
```diff
@@ -0,0 +1,209 @@
+If a single table needs to be written in a specific format that
+astropy does not support, one of the easiest solutions is to write a
+data file that is close to the desired format and then edit it by hand
+in a text editor. Often, we can add extra columns or place useful
+special values and then "search and replace all" in a text
+editor. Here, we want to write a table that says "end of line" at the
+end of every line. We can try to do that by simply adding another
+column to our table that has the content we are looking for::
```
This seems really contrived and I can't see this being "often" useful.

taldcroft commented on docs/io/ascii/howtoexamples.rst:
```diff
@@ -0,0 +1,209 @@
+  EXAMPLE END
+
+
+Write our own reader or writer class
+====================================
+
```
This is not a realistic option for the vast majority of users, having seen some recent work to add a new reader class, it pretty much requires help from the astropy devs. I think the existing https://docs.astropy.org/en/latest/io/ascii/read.html#advanced-customization section gets the point across.

taldcroft reviewed (changes requested):
@hamogu - there are some nuggets of useful advice here, but overall I'm suggesting descoping a bit to get this merged.

hamogu commented on docs/io/ascii/howtoexamples.rst:
```diff
@@ -0,0 +1,209 @@
+
+.. _io_ascii_how_to_examples:
+
+How to fix a table that is not read or written correctly
```
That's because I never finished what I indented to write here. This is supposed to get to e.g. https://github.com/astropy/astropy/pull/3812#issuecomment-314771680 which I think *is* useful - there is no guarantee that you and I will be on this project forever, so we should write it down. It can also be a path for others to become developers - that's how I got into astropy (with your help - remember the DAOPhotTableReader?)

hamogu commented on docs/io/ascii/howtoexamples.rst:
```diff
@@ -0,0 +1,209 @@
+
+Obtain the data table in a different format
+===========================================
+Sometimes we deal with legacy ACSII table and the data only exisits in this
+form. However, in other cases, we generate the ASCII table ourselves, for
+example, we export it from a spreadsheet or some other application. In that
+case, the easist solution is often to export it in a different format such
+as fits or as different ASCII form, such as fixed-width instead of
+comma-separated.
```
Fixed-width is quite a common format for tables written by legacy FORTRAN code, and sometimes C code, I;ve seen that in PHOENIX, FLASH, PLUTO, and a few others.

taldcroft replied:
The question was whether apps like FLASH both (1) emit broken output in one format (e.g. CSV), thus leading the user to this page; and (2) have an option to write data to an alternative format like fixed width or FITS that will not be broken. Generally speaking fixed width output from FORTRAN or C code is going to parse just fine.

hamogu commented:
I'll leave this on my list and will hopefully work in this and reopen when I have a downscoped version in the next few days.

hamogu commented on docs/io/ascii/howtoexamples.rst:
```diff
@@ -0,0 +1,209 @@
+  ... Vega A0
+  ... alpha Cen G2V+K1
+  ... '''
+  >>> ascii.read(table)
```
This is what I get in IPython or the Jupyter notebook without `guess=False` (copy and paste full output from the terminal. I think that's what most users use these days. We `guess=False`, I would not see "a long list of formats
that it tried on this table"

```
(kitchensink) MoritzAirRoseGold ~> ipython
Python 3.10.1 | packaged by conda-forge | (main, Dec 22 2021, 01:39:07) [Clang 11.1.0 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.30.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]:   >>> from astropy.io import ascii
   ...:   >>> table = '''
   ...:   ... Star SpT
   ...:   ... Vega A0
   ...:   ... alpha Cen G2V+K1
   ...:   ... '''
   ...:   >>> ascii.read(table)
---------------------------------------------------------------------------
InconsistentTableError                    Traceback (most recent call last)
~/mambaforge/envs/kitchensink/lib/python3.10/site-packages/astropy/io/ascii/ui.py in _guess(table, read_kwargs, format, fast_reader)
    521             reader = get_reader(**read_kwargs)
--> 522             dat = reader.read(table)
    523             _read_trace.append({'kwargs': copy.deepcopy(read_kwargs),

~/mambaforge/envs/kitchensink/lib/python3.10/site-packages/astropy/io/ascii/core.py in read(self, table)
   1368 
-> 1369                     raise InconsistentTableError(errmsg)
   1370 

InconsistentTableError: Number of header columns (2) inconsistent with data columns (3) at data line 1
Header values: ['Star', 'SpT']
Data values: ['alpha', 'Cen', 'G2V+K1']

During handling of the above exception, another exception occurred:

InconsistentTableError                    Traceback (most recent call last)
<ipython-input-1-732a26b21e0d> in <module>
      5 alpha Cen G2V+K1
      6 '''
----> 7 ascii.read(table)

~/mambaforge/envs/kitchensink/lib/python3.10/site-packages/astropy/io/ascii/ui.py in read(table, guess, **kwargs)
    339         # through below to the non-guess way so that any problems result in a
    340         # more useful traceback.
--> 341         dat = _guess(table, new_kwargs, format, fast_reader)
    342         if dat is None:
    343             guess = False

~/mambaforge/envs/kitchensink/lib/python3.10/site-packages/astropy/io/ascii/ui.py in _guess(table, read_kwargs, format, fast_reader)
    551                    '************************************************************************']
    552             lines.extend(msg)
--> 553             raise core.InconsistentTableError('\n'.join(lines))
    554 
    555 

InconsistentTableError: 
ERROR: Unable to guess table format with the guesses listed below:
Reader:Ecsv fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:FixedWidthTwoLine fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:RST fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:FastBasic fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Basic fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:FastRdb fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Rdb fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:FastTab fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Tab fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Cds fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Mrt fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Daophot fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:SExtractor fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Ipac fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Latex fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:AASTex fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:FastCommentedHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic fast_reader: {'enable': True} fill_values: [('', '0')]

************************************************************************
** ERROR: Unable to guess table format with the guesses listed above. **
**                                                                    **
** To figure out why the table did not read, use guess=False and      **
** fast_reader=False, along with any appropriate arguments to read(). **
** In particular specify the format and any known attributes like the **
** delimiter.                                                         **
************************************************************************

```

pllim commented:
This still looks stale, so I am remilestoning.

hamogu commented:
@taldcroft I pushed changes that hopefully address your comments a while ago. I guess you were waiting for me to fix the remaining sphinx issues and I was waiting for you to broadly say "looks like the right direction" before I dot the i's and cross the t's. 
The previous logs are not available any longer, so I rebased and pushed again, but please let me know how you want to proceed - should I make sure all tests come back green before you look at it or do you have time to look first so I don't waste time fixing um formatting in sections that you later ask me to remove anyway?

taldcroft commented:
@hamogu - this just fell off my radar. I'll have a look.

taldcroft commented:
> @taldcroft I pushed changes that hopefully address your comments a while ago.

@hamogu - I don't see any sign of changes since my comments. Normally GitHub would be showing `Outdated` where changes have been made and I would see applicable commits. Looking again at the current file diffs, I can verify that none of the comments I made have been addressed (ranging from big picture stuff like delete this whole section down to fixing typos).

We might want to get review from someone else because I think we have pretty different visions on this and maybe mine is not the best. For instance @mhvk gave this a :+1: and @dhomeier is a maintainer and @bsipocz has commented on `io.ascii` issues.

hamogu commented:
Mmm. I thought I had addressed them and GH did show me an "outdated", but then I rebased and force pushed. So, I probably forces-pushed from a branch that did not have the newest version and overwrite the changes that I had already made. I'll fix that and then ping you again.

hamogu commented on docs/io/ascii/howtoexamples.rst:
```diff
@@ -0,0 +1,209 @@
+deal with tables that fail to read or cannot be written with the
+default settings.
+
+Obtain the data table in a different format
```
When I encounter a table that does not read, trying `guess=False` is usually not the first thing I do. I usually try to get a different format (fits, csv instead of space-separated etc.) first because then I don't have ot find out what the options are that allow me to read it with the ascii readers.

hamogu commented:
@taldcroft Please look again. I see a lot of "outdated" now. I will split off the content from https://github.com/astropy/astropy/pull/3812#issuecomment-314797798 in the separate PR since that's a lot more advanced and we might agree on having this PR, but not the other one.

taldcroft commented on docs/io/ascii/read.rst:
```diff
@@ -859,3 +861,125 @@ them at the end.
+The purpose of this section is to provide a few examples how we can
+deal with tables that fail to read.
+
+Obtain the data table in a different format
```
```suggestion
Obtain the Data Table in a Different Format
```

taldcroft commented on docs/io/ascii/read.rst:
```diff
@@ -859,3 +861,125 @@ them at the end.
+Making a table easier to read
+-----------------------------
+
+Sometimes, the parameters for `astropy.io.ascii.read` to specify e.g.
```
(See https://docs.astropy.org/en/latest/development/style-guide.html#abbreviations)
```suggestion
Sometimes, the parameters for `astropy.io.ascii.read` to specify, for example
```

taldcroft commented on docs/io/ascii/read.rst:
```diff
@@ -859,3 +861,125 @@ them at the end.
+
+
+Badly formatted header line
+---------------------------
```
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

taldcroft commented on docs/io/ascii/read.rst:
```diff
@@ -859,3 +861,125 @@ them at the end.
+  EXAMPLE END
+
+Badly formatted data line
+-------------------------
```
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^
```

taldcroft commented on docs/io/ascii/read.rst:
```diff
@@ -859,3 +861,125 @@ them at the end.
+after the header lines and counting the data lines from 0 as usual in Python). In this table with
+just two lines the problem is easy to spot, but for longer tables, the line number is
+very helpful. We can now fix that line by hand in the file by adding quotes
+around ``"alpha Cen"``. Then, we can try to read the table again and see
```
```suggestion
around ``"alpha Cen"``. Then we can try to read the table again and see
```

taldcroft commented on docs/io/ascii/read.rst:
```diff
@@ -859,3 +861,125 @@ them at the end.
+
+.. _io_ascii_how_to_examples:
+
+How to fix a table that is not read correctly
+=============================================
```
```suggestion
How to Find and Fix Problems Reading a Table
============================================
```

taldcroft reviewed (changes requested):
Thanks @hamogu - basically this looks good now, I just have a bunch of formatting / style suggestions. There is one substantial suggestion to expand on finding the problem with an example and explaining things a bit more.

hamogu commented:
I agree with all suggested changes and merged them in. Thanks for reviewing! Now, I'll just need to check if there are any formatting problems that the CI flags.

pllim commented:
Just some trailing whitespace. 😉 

```
trim trailing whitespace.................................................Failed
- hook id: trailing-whitespace
- duration: 0.18s
- exit code: 1
- files were modified by this hook
```

pllim commented on docs/io/ascii/read.rst:
```diff
@@ -859,3 +861,155 @@ them at the end.
+
+In order to find the actual problem with the reading this file, you would do::
+
+  >>> ascii.read(txt, format='commented_header', delimiter=' ', guess=False, fast_reader=False)
```
You might need ellipses (sp?) directive for the doctest, if this is doctested, but let's wait for the CI to complain first after you fix the trailing whitespace.

p.s. You can set your IDE to auto-remove trailing whitespace, I think.

taldcroft reviewed (approved):
Thanks @hamogu !!
